### PR TITLE
Attachment links

### PR DESCRIPTION
**Pull request by brookeclouston**
This branch addresses the RemoveLinksToAttachments script. It is in response to [Redmine Support #4173 ](https://redmine.biodiversity.agr.gc.ca/issues/4173)

### CODE REVIEW

File: RemoveLinksToAttachments.py [hunk 1]
```diff
@@ -2,11 +2,14 @@
 Redmine Support #4173
```
Comment (kcranston) [inline RemoveLinksToAttachments.py]: remove the link to the redmine support ticket from the script itself

File: RemoveLinksToAttachments.py [hunk 2]
```diff
@@ -2,11 +2,14 @@
...
 Removes all attachments and references/links to attachments from the schema by selecting ID's from collection objects
```
Comment (kcranston) [inline RemoveLinksToAttachments.py]: description should make it clear that the script removes attachments where the MIME type is `image/jpeg`, not _all_ attachments.

File: RemoveLinksToAttachments.py [hunk 3]
```diff
@@ -22,69 +25,54 @@ def select_references(db):
                          " AND REFERENCED_COLUMN_NAME LIKE 'AttachmentID' and table_name != 'attachment'")
     return db_reference.fetchall()
 
-# writes the data passed in to a csv file
-def save_report(file_name,data):
-    write_report(file_name,["Attachment ID"],data)
-    print("Report saved as '%s.csv'" % file_name)
-
-# deletes any entry that references the attachmentID, then deletes attachmentID from the schema, handels any conflicts
```
Comment (kcranston) [inline RemoveLinksToAttachments.py]: handels -> handles

File: RemoveLinksToAttachments.py [hunk 4]
```diff
@@ -22,69 +25,54 @@ def select_references(db):
...
-    if display == "show":
-        for c in conflicts:
-            print(c)
-    elif display == "save":
-        save_report("AttachmentConflicts",conflicts)
-    elif display == "n":
-        pass
-    else:
-        print("Invalid command")
+        for table in select_references(db):
```
Comment (kcranston) [inline RemoveLinksToAttachments.py]: why did you remove the exception handling here? How does the user know if one or more of the deletes fail?

File: RemoveLinksToAttachments.py (hunk 1 above)
Comment (brookeclouston) [reply RemoveLinksToAttachments.py]: Banner has been rewritten to remove the redmine support ticket

File: RemoveLinksToAttachments.py (hunk 2 above)
Comment (brookeclouston) [reply RemoveLinksToAttachments.py]: Banner has been rewritten to make it clear that the script removes the attachments with MimeType as image/jpegs
https://github.com/AAFC-BICoE/specify-scripts/blob/f1c1ce7bae43fde6c4cef3543d95ead5bdbd412d/RemoveLinksToAttachments.py#L2

File: RemoveLinksToAttachments.py (hunk 4 above)
Comment (brookeclouston) [reply RemoveLinksToAttachments.py]: Script has been reformatted to include exception handling 
https://github.com/AAFC-BICoE/specify-scripts/blob/f1c1ce7bae43fde6c4cef3543d95ead5bdbd412d/RemoveLinksToAttachments.py#L29-L38

File: RemoveLinksToAttachments.py [hunk 5]
```diff
@@ -22,69 +22,65 @@ def select_references(db):
...
     db_delete_attachment = db.cursor()
     conflicts = []
     for record in attachments:
         try:
-            for table in select_references(db):
+            for table in referenced_tables:
                 db_delete_reference.execute("DELETE FROM %s WHERE AttachmentID = %s" % (table[0], record[0]))
             db_delete_attachment.execute("DELETE FROM attachment WHERE AttachmentID = %s" % record[0])
             db.commit()
         except:
```
Comment (kcranston) [inline RemoveLinksToAttachments.py]: This is awfully generic exception handling. What sort of conflict can happen? Is that more than one possible way that this code might fail? There is a loop with two actions inside the try block - what if some, but not all, fail? Should there be a rollback in case of problems?

File: RemoveLinksToAttachments.py [hunk 6]
```diff
@@ -22,69 +22,65 @@ def select_references(db):
...
+        file_name = "AttachmentsToDelete[%s]" % (datetime.date.today())
+        heading = ["Attachment ID"]
+        write_report(file_name, heading, attachments)
+        if not delete:
+            print("No changes made, report of %s attachmentIDs to be deleted saved as %s.csv" % (len(attachments),file_name))
+            return print("See -h for more options")
+        return print("Report saved as '%s.csv'" % file_name)
 
 if __name__ == "__main__":
     main()
```
Comment (kcranston) [inline RemoveLinksToAttachments.py]: should be a newline at end of file

**Comment by kcranston** [comment]
Can you move all of the test scripts into their own `tests` subdirectory?

File: RemoveLinksToAttachments.py (hunk 5 above)
Comment (brookeclouston) [reply RemoveLinksToAttachments.py]: Better exception handling has been added. The only conflict that can really happen is a foreign key constraint failing, in which case MySQL wont let the entry be deleted. In the event of this conflict occurring in either the first step (removing the references) or the second (deleting the entry), the loop moves onto the next attachmentID (so neither is deleted), the conflicting AttachmentID is added to a conflict list and once the loop has finished a conflict report is created, the number of conflicts and number of successful entries deleted are shown.

File: RemoveLinksToAttachments.py (hunk 6 above)
Comment (brookeclouston) [reply RemoveLinksToAttachments.py]: Newline has been added

**Comment by brookeclouston** [comment]
All tests have been moved to subdirectory 'tests'.

File: tests/test_RemoveLinksToAttachments.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
+"""
+Tests the RemoveLinksToAttachments script for 4 different cases. First creates a test database with 4 tables, populated
+with test data to be passed into the original script. Tests are preformed to confirm that attachmentID's with mimeType
+'image/jpeg' are selected properly, that attachmentID's that create no conflicts are removed properly, that when the
+function is called with no conflicts an empty conflict list is returned, and that when the function is called with
+conflicts a non empty conflict list is returned
+"""
+import RemoveLinksToAttachments, os, test_db, sqlite3, unittest
```
Comment (kcranston) [inline tests/test_RemoveLinksToAttachments.py]: what is the test_db module? does this get used anywhere?

File: tests/test_RemoveLinksToAttachments.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        test_referenced_tables = [("testdata1",),("testdata2",),("collectionobjectattachment",)]
+        test_attachments = [("1",),("2",),("3",)]
+        actual = RemoveLinksToAttachments.delete_attachments (conn, test_referenced_tables,test_attachments)
+        expected = []
+        self.assertEqual(expected,actual)
+
+    def test_delete_attachments_conflicts(self):
+        """ tests to confirm that when attachments are deleted and conflicts occur, corrected values are returned"""
+        conn = sqlite3.connect("specifytest.db")
+        test_referenced_tables = [("testdata1",),("testdata",),("collectionobjectattachment",)]
```
Comment (kcranston) [inline tests/test_RemoveLinksToAttachments.py]: `sqlite3.OperationalError: no such table: testdata` - did you forget to create this table, or is this a typo in the table name?

**Comment by kcranston** [comment]
I'd like to get this one merged, but I cannot run the tests - see comments above. 

(NB: I removed the irrelevant script and committed it to another branch)

File: tests/test_RemoveLinksToAttachments.py (hunk 7 above)
Comment (brookeclouston) [reply tests/test_RemoveLinksToAttachments.py]: the test_db does not get used and has been removed

File: tests/test_RemoveLinksToAttachments.py (hunk 8 above)
Comment (brookeclouston) [reply tests/test_RemoveLinksToAttachments.py]: Test script has been updated so there is no longer an error thrown

**Comment by brookeclouston** [comment]
Both the original script and the test script have been updated so they should be ready to be merged.

**Comment by kcranston** [comment]
I moved moved the files that aren't about attachments to another branch, and moved the test file back to the parent (because you can't currently run the tests from the subdirectory). Let's fix the repository structure in a different PR.
